README.md
# NotePlan Tools

**Note: Much of the capability of this script is now available through the Plugins I've written that are available through NotePlan v3.  Please see [this overview](https://help.noteplan.co/article/65-commandbar-plugins), and the individual [plugins' details in the GitHub repository](https://github.com/NotePlan/plugins). In particular see 'Repeat Extensions' and 'Event Helpers'.**

`npTools.rb` is a Ruby script that adds functionality to the [NotePlan app](https://noteplan.co/). Particularly when run frequently, this provides a more flexible system for repeating tasks, allows for due dates to be expressed as offsets and therefore templates, moves items from Daily files to Note files, and creates events. It incorporates an earlier script to 'clean' or tidy up NotePlan's data files.

Each time the script runs, it does a number of things, explained in each section:

### Tidy up

It **tidies up** data files, by:
1. removing the time part of any `@done(...)` mentions that NotePlan automatically adds when the 'Append Completion Date' option is on.
2. removing any `@done(...)` mentions on completed Checklist items that NotePlan automatically adds when the 'Append Completion Date' option is on. (From NP v3.8)
3. removing `#waiting` or `#high` tags or `<dates` from completed or cancelled tasks (configurable)
4. removing scheduled (`* [>] task`) items in calendar files (as they've been copied to a new day)
5. removing any lines with just `* `or `-` or starting `#`s
6. removing header lines without any content before the next header line of the same or higher level (i.e. fewer `#`s)
7. removing any multiple consecutive blank lines.

### Moves Daily (Calendar) note items
(From v1.9.2) The script can now move any Daily note entries with a **`>date`** in it to the mentioned Daily note. To do so requires turning on through the `-m` option.

In more detail:
- where the line is a heading, it moves the heading and all following lines until a blank line, or the next heading of the same level
- where the line isn't a heading, it moves the line and any following indented lines (optionally terminated by a blank line)
- the lines are inserted after a section heading (e.g. '### Tasks') as defined in the DAILY_TASKS_SECTION_NAME constant (or after header if this is blank). (To configure this constant, see below.)

NB: This only operates from Daily (Calendar) notes; it therefore doesn't interfere with **linking and back-linking** between main notes.

### Files Daily (Calendar) note tasks
The script can move any Daily note tasks with **a `[[Note title]]`** in it to the mentioned note, **filing** them directly after the header section.

In more detail:
- where the line is a heading, it moves the heading and all following lines until a blank line, or the next heading of the same level
- where the line isn't a heading, it moves the line and any following indented lines (optionally terminated by a blank line)
- if there is a `>YYYY-MM-DD` date specified in the line already, then carry that over, otherwise  add today's date

This feature can be turned on using two different options, with slightly different triggers:
- `-m` (`--move`): move tasks with such a note link, _whether or not the task is complete_;
- `-t` (`--movecomplete`): move only _completed tasks_.

NB: This only operates from Daily (Calendar) notes; therefore it _doesn't_ interfere with linking and back-linking between regular notes.

### Templates for dates
It changes any mentions of **date offset patterns** (such as `{-10d}`, `{+2w}`, `{-3m}`) into scheduled dates (e.g. `>2020-02-27`), if it can find a valid date pattern in the previous heading, previous main task if it has sub-tasks, or in the line itself. This allows for users to define simple **template sections** and copy and paste them into the note, set the due date at the start, and the other dates are then worked out for you.

| For example ...                                                                                                                                                                        | ... becomes                                                                                                                                                                                                 |
| --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| \#\#\# Christmas Cards 25/12/2020<br />\* Write cards {-20d}<br />\* Post overseas cards {-15d}<br />\* Post cards to this country {-10d}<br />\* Store spare cards for next year {+3d} | \#\#\# Christmas Cards 25/12/2020<br />\* Write cards >2020-12-05<br />\* Post overseas cards >2020-12-10<br />* Post cards to this country >2020-12-15<br />\* Store spare cards for next year >2020-12-28 |
| \* Bob's birthday on 14/09/2020<br />&nbsp;&nbsp;\* Find present {-6d}<br />&nbsp;&nbsp;\* Wrap & post present {-3d} <br />&nbsp;&nbsp;\* Call Bob {0d}                                 | \* Bob's birthday on 14/09/2020<br />&nbsp;&nbsp;\* Find present >2020-09-08<br />&nbsp;&nbsp;\* Wrap & post present >2020-09-11<br />&nbsp;&nbsp;\* Call Bob >2020-09-14                                   |

You can use this within a line to have both a **deadline** and a calculated **start date**:

| For example ...                                                      | ... becomes                                                                       |
| ---------------------------------------------------------- | -------------------------------------------------------------------- |
| * Post cards deadline 2020-12-18 {-10d} | * Post cards deadline 2020-12-18 >2020-12-08 |

In more detail:

- Valid **date offsets** are specified as `[+][0-9][bdwmqy]`. This allows for `b`usiness days,  `d`ays, `w`eeks, `m`onths, `q`uarters or `y`ears. (Business days skip weekends. If the existing date happens to be on a weekend, it's treated as being the next working day. Public holidays aren't accounted for.)  There's also the special case `{0d}` meaning on the day itself.
- It ignores offsets in a section with a heading that includes a `#template` hashtag
- The base date is by default of the form `YYYY-MM-DD`, not preceded by characters `0-9(<>`, all of which could confuse.
- But you can add another custom date format to use by setting the `RE_DATE_FORMAT_CUSTOM` variable (see below). The dates matched by this regular expression must additionally be caapble of being correctly parsed by ruby's Date.parse() built-in command, over which I have no control.

### Create new Calendar Events
NotePlan allows for time-blocking, but this only works to its in-built calendar display. With npTools you also **create events** that are visible and editable in Apple Calendar, or any other apps that use iCloud, *not just NotePlan*.  To do this use the `#create_event` tag on a line (task, comment or heading) with a timeblocking command such as `3:00-3:45[AM|PM]` or `3-5pm` or `3PM` and optional location such as `at Jim's`. This allows for meeting events to be listed on a day, and also created in the calendar. In combination with the date offset patterns above, it further allows scheduling preparation time (for example) days or hours before events.

The date of the event is determined in this order:
1. it will use a `>YYYY-MM-DD` mentioned in the line. NB:  date offset patterns get calculated before it looks to create events.
2. if it's a calendar note, then use that date
3. otherwise use today's date.

For example to create a meeting event for a certain day put this in it's daily (calendar) note:
```
### Project X Meeting #create_event 10:30am at Jim's
```

To extend this and use the date templates, add a timed task with calendar entry to do some associated tasks 5 days before it, and the following morning:
```
### Project X Meeting 21-12-2020 #create_event 10:30am at Jim's
* write and circulate agenda {-5d} #create_event 4pm
* send out actions {1d} #create_event 9am
```

![Video showing event creation](npTools-create_event.gif)

Notes on this:
- If no date is given on the line, and 
- If no finish time is set, then the event defaults to an hour.
- You can use the shortcut `3PM` or `3-5PM` when you don't need to specify the minutes -- though NotePlan itself currently doesn't recognise this syntax for time blocks. You can specify minutes as well either using `3:45` or `3.15` formats.
- If am/AM or pm/PM isn't given, then the hours are assumed to be in 24-hour clock
- It's best to put any `at place` location at the end of the line, as there's no easy way of telling how a location finishes, so it will use the rest of the line as the location.
- Any indented following lines are copied to the event description, with leading whitespace removed.
- Under the hood this uses AppleScript, and takes a few seconds per event.  Once it has run succesfully the `#create_event` is changed to `#event_created` so that it won't be triggered again.  
- There are some calendar settings that need to be configured for this: see Installation and Configuration below.

### Extend the existing @repeat mechanism
It **creates new repeats** for newly completed tasks that include a `@repeat(interval)`, on the appropriate future date.

- Valid intervals are specified as `[+][0-9][bdwmqy]`. This allows for `b`usiness days, `d`ays, `w`eeks, `m`onths, `q`uarters or `y`ears.
- When _interval_ is of the form `+2w` it will duplicate the task for 2 weeks after the date the _task was completed_.
- When _interval_ is of the form `2w` it will duplicate the task for 2 weeks after the date the _task was last due_. If this can't be determined, then it defaults to the first option.

NB: For this feature to work, you need to have the 'Append Completion Date'  NotePlan setting turned on, and to have the first type of tidy up (above) happening.

<!-- In future, extending the **archiving** system. -->

## Running the Tools
There are 2 ways of running the script:
1. with no arguments (`ruby npTools.rb`), it checks all note and daily files updated in the last 24 hours. This is the way to use it automatically, running one or more times each day. (This is configurable by `HOURS_TO_PROCESS` below.)
2. with passed filename pattern(s), where it works on any matching Calendar or Note files. For example, to match the Daily file from 24/3/2020 use `ruby npTools.rb 20200324.txt`. It can include wildcard *patterns* to match multiple files, for example `"202003*.txt"` to process all Daily files from March 2020.  (It now needs to be in double quotes for the file pattern matching to work.) If no `.` is found in the pattern, the pattern matches all files as `"*pattern*.*"`.

You can also specify the following **options**:
- `-h` (`--help`) for a list of options, 

CHANGELOG.md
# CHANGELOG
As always, `npTools -h` to see the full list of current options.

## v2.4.1, 2023-01-26
- [Added] Support for moving lines when its checklist is completed (extending `-t` feature).

## v2.4.0, 2022-12-31
- [Added] function to remove `@done(...)` markers from completed checklist items (introduced in NotePlan 3.8). This is triggered by new command line option `-r`.

## v2.3.1, 2022-10-03
- [Fixed] bug with improperly closed frontmatter

## v2.3.0, 2022-06-26
- [Added] supports new Weekly notes, available with NotePlan 3.6.0.

## v2.2.2, 2022-05-17
- [Change] Removed the code to make new notes where linked notes aren't found, as it was unreliable. It will now simply warn the user.

## v2.2.0, 2022-05-10
- [Add] Adds the `--movecomplete` option to move a task from a calendar note to the project note indicated by [[note#title]] but only when the task has been completed. (The earlier '--move' option remains; this does the same, but operates whether or not the task has been completed.)
- [Add] Understands notes with titles in the frontmatter (added NP v3.4.x)
- [Change] Removed `--archive` option, as it isn't ready for use. (It was already turned off by default.)
- [Change] Tidy up logging

## v2.1.0, 31.10.2021
- [Change] The @repeat(..) function now creates the next copy of the repeated task without using the scheduled indicator `[>]` but the regular open task indicator `[ ]`.

## v2.0.0, 31.10.2021
- [Change] Change to using `#!/usr/bin/env ruby` at the start of scripts to make it easier to pick up whatever ruby installation is the user's preference.

## v1.9.7, 28.2.2021
- [Improve] The place where lines moved from daily notes are inserted is now smarter still.
- [Fix] Bugs introduced in refactoring regexes in last release.

## v1.9.4, 23.1.2021
- [Improve] The place where lines moved from daily notes are inserted is now smarter [Issue 39]. Note this removes the NUM_HEADER_LINES user-settable variable.

## v1.9.3, 23.1.2021
- [Fix] Fixed cosmetic bug in moving titled section from daily notes to a project note
- [Fix] Removed stray `>dates` when creating new `@repeat(...)`s

## v1.9.2, 14.2.2021
- [New] Now allow for `>date` events to be moved to from one daily to the one it points to. This has to be activated with the `-m` option, as it will be quite a significant change for some users.

## v1.9.1, 18.1.2021
- [Fix] Fixed bug in file-matching logic [thanks to Dimitry for reporting]
- [Fix] Fixed bug that stopped date pattern matching in calendar files [thanks to Dimitry for reporting]
- [Improve] Clarified the README for how dates are defined when using #create_event.

## v1.9.0, 16.1.2021
- [Add] Command line option `-c` to override how many hours to look back to find changed notes to process
- [Improve] Allow more types of time spec when creating events (e.g. "2.45PM" or "3.15-5.00"
- [Improve] For template dates be more discriminating about where to find dates to match on, so now ignores dates embedded in certain URLs.
- [Fix] spacing around processed template dates (e.g. {3d})
- [Fix] removing headers with empty sections now won't remove header if next content is a lower-level header
- [Clean up] Remove some obsolete code and add some more logging

## v1.8.6 8.1.2021
- [Fix] Allow event creation to work with "Language & Region" settings that use a 12-hour not a 24-hour clock [Issue 37]

## v1.8.6 8.1.2021
- [Add] Works with tasks that use the `- ` and `- [ ]` markers, as well as `* ` and `* [ ]` [Issue 24]

## v1.8.5 8.1.2021
- [Add] For event creation can now also specify time patterns of form 4-6PM and just 3-4
- [Add] Extend logging to see why some files aren't matching for a user

## v1.8.4 8.1.2021
- [Add] Add a second custom date match style for base dates to use in date offset patterns. See README for RE_DATE_OFFSET_CUSTOM.
- [Change] The built-in date match style for base dates is now NotePlan's usual YYYY-MM-DD, not DD-MM-YYYY etc.

## v1.8.3 6.1.2021
- [Improve] Can now use simpler '3PM' type of time spec when creating events.

## v1.8.2. 2.1.2021
- [Improve] Can now customise the `#create_event` tag used to trigger creating events. Also opens the selected Calendar app if needed first, and reduces the delay if not. [Issue 36]

## v1.8.1. 23.12.2020
- [Improve] Can now add location to created events, and copies any description from following indented lines. [Issue 36]

## v1.8.0. 21.12.2020
- [New] Add ability to create events in the Calendar, based on time-blocking syntax. See README for more details. [Issue 36]

## v1.7.4. 19.12.2020
- [Improve] Clarify documentation to show all sorts of lines in daily notes are moved to the mentioned [[note title]], not just task or header lines. Also loosened unnecessarily strict regexes used here.

## v1.7.3. 10.12.2020
- [New] Allow use of weekdays in repeats and template dates (using 'b' rather than usual 'd' for days) [Issue 32]

## v1.7.2, 5.12.2020
- [Improve] Extended the command line option --skiptoday to allow comma-separated list of notes to ignore [thanks to @BMStroh PR32]

## 1.7.1, 26.11.2020
- [New] add --skipfile=file option to ignore particular files [thanks to @BMStroh, issue 30]
- [Fix] Blank headers at EOF not removed [thanks to @BMStroh, PR31]

## 1.7.0, 19.11.2020
- [New] where the note for a  `[[Note link]]` doesn't exist, it is created in the top-level Notes folder first

## v1.6.1, 13.11.2020
- [New] remove [>] tasks from calendar notes, as there will be a duplicate (whether or not the 'Append links when scheduling' option is set or not)


npMediaSave.rb
#!/usr/bin/env ruby
#-------------------------------------------------------------------------------
# Script to Save some Media notes into NotePlan
# by Jonathan Clark, v0.3.3, 20.3.2021
#
# v0.3 now copes with multi-line tweets
#-------------------------------------------------------------------------------
VERSION = "0.3.3"
require 'date'
require 'cgi'
require 'colorize'
require 'optparse' # more details at https://docs.ruby-lang.org/en/2.1.0/OptionParser.html

#-------------------------------------------------------------------------------
# Setting variables to tweak
#-------------------------------------------------------------------------------
MEDIA_STRING = '### Media' # the title of the section heading to add these notes to
NOTE_EXT = "md" # or "txt"
IFTTT_FILEPATH = "/Users/jonathan/Dropbox/IFTTT/"
IFTTT_ARCHIVE_FILEPATH = "/Users/jonathan/Dropbox/IFTTT/Archive/"
INSTAPAPER_FILE = "Instapaper Archived Items.txt"
MEDIUM_FILE = "Medium Articles.txt"
SPOTIFY_FILE = "Spotify Saved Tracks.txt"
TWITTER_FILE = "My Tweets.txt"
YOUTUBE_LIKES_FILE = "YouTube liked videos.txt"
YOUTUBE_UPLOAD_FILE = "YouTube upload.txt"
DATE_TIME_LOG_FORMAT = '%e %b %Y %H:%M'.freeze # only used in logging
DATE_TIME_APPEND_FORMAT = '%Y%m%d%H%M'.freeze

#-------------------------------------------------------------------------------
# To use test data instead of live data, uncomment relevant definitions.
# NB: assumes one per line, apart from Twitter.
#-------------------------------------------------------------------------------
# $spotify_test_data = <<-END_S_DATA
# February 6, 2021 at 11:11PM | Espen Eriksen Trio | In the Mountains | Never Ending January | https://ift.tt/2TRqQiB | https://ift.tt/2LptJng
# February 6, 2021 at 11:56AM | Brian Doerksen | Creation Calls | Today | https://ift.tt/2qQI2Sq | https://ift.tt/3pYs1by
# END_S_DATA

# $instapaper_test_data = <<-END_I_DATA
# February 6, 2021 at 05:49AM \\ Thomas Creedy: Imago Dei \\ https://ift.tt/3rJl1Bh \\
# February 6, 2021 at 06:02AM \\ Is the 'seal of the confessional' Anglican? \\ https://ift.tt/2MpMAPX \\ "Andrew Atherstone writes: The Church of England has at last published the report of the 'Seal of the Confessional' working party , more than a year after it..."
# February 6, 2021 at 04:04PM \\ In what ways can we form useful relationships between notes? \\ https://ift.tt/3aT3LS3 \\ "Nick Milo Aug 8, 2020 * 7 min read Are you into personal knowledge management PKM)? Are you confused about when to use a folder versus a tag versus a link..."
# END_I_DATA

# $twitter_test_data = <<-END_T_DATA
# February 6, 2021 at 03:56PM | A useful thread which highlights the problems of asking the wrong question. https://t.co/xEKimf9cLK | jgctweets | http://twitter.com/jgctweets/status/1353371090609909760
# END_T_DATA

#-------------------------------------------------------------------------------
# Other Constants & Settings
#-------------------------------------------------------------------------------
USERNAME = ENV['LOGNAME'] # pull username from environment
USER_DIR = ENV['HOME'] # pull home directory from environment
DROPBOX_DIR = "#{USER_DIR}/Dropbox/Apps/NotePlan/Documents".freeze
ICLOUDDRIVE_DIR = "#{USER_DIR}/Library/Mobile Documents/iCloud~co~noteplan~NotePlan/Documents".freeze
CLOUDKIT_DIR = "#{USER_DIR}/Library/Containers/co.noteplan.NotePlan3/Data/Library/Application Support/co.noteplan.NotePlan3".freeze
# TodaysDate = Date.today # can't work out why this needs to be a 'constant' to work -- something about visibility, I suppose
np_base_dir = CLOUDKIT_DIR if Dir.exist?(CLOUDKIT_DIR) && Dir[File.join(CLOUDKIT_DIR, '**', '*')].count { |file| File.file?(file) } > 1
NP_CALENDAR_DIR = "#{np_base_dir}/Calendar".freeze

# Colours to use with the colorization gem
# to show some possible combinations, run  String.color_samples
# to show list of possible modes, run   puts String.modes  (e.g. underline, bold, blink)
CompletedColour = :light_green
InfoColour = :yellow
ErrorColour = :light_red
# Test to see if we're running interactively or in a batch mode:
# if batch mode then disable colorisation which doesn't work in logs
tty_code = `tty`.chomp
String.disable_colorization true if tty_code == 'not a tty'

# Variables that need to be globally available
time_now = Time.now
$date_time_now_log_fmttd = time_now.strftime(DATE_TIME_LOG_FORMAT)
$date_time_now_file_fmttd = time_now.strftime(DATE_TIME_APPEND_FORMAT)
$verbose = false
$npfile_count = 0


#-------------------------------------------------------------------------
# Helper Functions
#-------------------------------------------------------------------------
def main_message_screen(message)
  puts message.colorize(CompletedColour)
end

def warning_message_screen(message)
  puts message.colorize(InfoColour)
end

def error_message_screen(message)
  puts message.colorize(ErrorColour)
end

def log_message_screen(message)
  puts message if $verbose
end

#-------------------------------------------------------------------------
# Class definition: NPCalFile
#-------------------------------------------------------------------------
class NPCalFile
  # Define the attributes that need to be visible outside the class instances
  attr_reader :id, :media_header_line, :is_calendar, :is_updated, :filename, :line_count

  def initialize(date)
    # Create NPFile object from reading Calendar file of date YYYMMDD

    # Set the file's id
    $npfile_count += 1
    @id = $npfile_count
    @filename = "#{NP_CALENDAR_DIR}/#{date}.#{NOTE_EXT}"
    @lines = []
    @line_count = 0
    @title = date
    @is_updated = false

    begin
      log_message_screen(" Reading NPCalFile for '#{@title}'")

      # Open file and read in all lines (finding any Done and Cancelled headers)
      # NB: needs the encoding line when run from launchctl, otherwise you get US-ASCII invalid byte errors (basically the 'locale' settings are different)
      f = File.open(@filename, 'r', encoding: 'utf-8')
      n = 0
      f.each_line do |line|
        @lines[n] = line
        n += 1
      end
      f.close
      @line_count = @lines.size # e.g. for lines 0-2 size => 3
      log_message_screen(" -> Read NPCalFile for '#{@title}' using id #{@id}. #{@line_count} lines")
    rescue StandardError => e
      error_message_screen("ERROR: #{e.exception.message} when re-writing note file #{@filename}")
    end
  end

  def insert_new_line(new_line, line_number)
    # Insert 'line' into position 'line_number'
    # NB: this is insertion at the line number, so that current line gets moved to be one later
    n = @line_count # start iterating from the end of the array
    line_number = n if line_number >= n # don't go beyond current size of @lines
    log_message_screen("  insert_new_line at #{line_number} (count=#{n}) ...")
    @lines.insert(line_number, new_line)
    @line_count = @lines.size
  end

  def append_line_to_section(new_line, section_heading)
    # Append new_line after 'section_heading' line.
    # If not found, then add 'section_heading' to the end first
    log_message_screen("  append_line_to_section for '#{section_heading}' ...")
    n = 0
    added = false
    found_section = false
    while !added && (n < @line_count)
      line = @lines[n].chomp
      # if an empty line or a new header section starting, insert line here
      if found_section && (line.empty? || line =~ /^#+\s/)
        insert_new_line(new_line, n)
        added = true
      end
      # if this is the section header of interest, save its details. (Needs to come after previous test.)
      found_section = true if line =~ /^#{section_heading}/
      n += 1
    end
    # log_message_screen("  section heading not found, so adding at line #{n}, #{@line_count}")
    insert_new_line(section_heading, n) unless found_section # if section not yet found then add it before this line
    insert_new_line(new_line, n + 1) unless added # if not added so far, then now append
  end

  def rewrite_cal_file
    # write out this updated calendar file
    main_message_screen("  > writing updated version of #{@filename}")
    # open file and write all the lines out
    begin
      File.open(@filename, 'w') do |f|
        @lines.each do |line|
          f.puts line
        end
      end
    rescue StandardError => e
      error_message_screen("ERROR: #{e.exception.message} when re-writing calendar file #{filepath}")
    end
  end
end

#--------------------------------------------------------------------------------------
# SPOTIFY
#--------------------------------------------------------------------------------------
def process_spotify
  spotify_filepath = IFTTT_FILEPATH + SPOTIFY_FILE
  catch (:done) do  # provide a clean way out of this
    if defined?($spotify_test_data)
      f = $spotify_test_data
      log_message_screen("Using Spotify test data")
    elsif File.exist?(spotify_filepath)
      if File.empty?(spotify_filepath)
        warning_message_screen("Spotify file empty")
        throw :done
      else
        f = File.open(spotify_filepath, 'r', encoding: 'utf-8')
      end
    else
      warning_message_screen("No Spotify file found")
      throw :done
    end

    begin
      f.each_line do |line|
        # Parse each line
        parts = line.split('|')
        log_message_screen(parts)
        # parse the given date-time string, then create YYYYMMDD version of it
        date_YYYYMMDD = Date.parse(parts[0]).strftime('%Y%m%d')
        log_message_screen("  Found item to save with date #{date_YYYYMMDD}:")

        # Format line to add
        line_to_add = "- new #spotify fave **#{parts[1].strip}**'s [#{parts[2].strip}](#{parts[4].strip}) from album **#{parts[3].strip}** ![album art](#{parts[5].strip})"
        log_message_screen(line_to_add)

        # Read in the NP Calendar file for this date
        this_note = NPCalFile.new(date_YYYYMMDD)

        # Add new lines to end of file, creating a ### Media section before it if it doesn't exist
        this_note.append_line_to_section(line_to_add, MEDIA_STRING)
        this_note.rewrite_cal_file
        main_message_screen("-> Saved new Spotify fave to #{date_YYYYMMDD}")
      end

      unless defined?($spotify_test_data)
        f.close
        # Now rename file to same as above but _YYYYMMDDHHMM on the end
        archive_filename = "#{IFTTT_ARCHIVE_FILEPATH}#{SPOTIFY_FILE[0..-5]}_#{$date_time_now_file_fmttd}.txt"
        File.rename(spotify_filepath, archive_filename)
      end
    rescue StandardError => e
      error_message_screen("ERROR: #{e.exception.message} for file #{SPOTIFY_FILE}")
    end
  end
end

#--------------------------------------------------------------------------------------
# INSTAPAPER
#--------------------------------------------------------------------------------------
def process_instapaper
  instapaper_filepath = IFTTT_FILEPATH + INSTAPAPER_FILE
  catch (:done) do  # provide a clean way out of this
    if defined?($instapaper_test_data)
      f = $instapaper_test_data
      log_message_screen("Using Instapaper test data")
    elsif File.exist?(instapaper_filepath)
      if File.empty?(instapaper_filepath)
        warning_message_screen("Note: Instapaper file empty")
        throw :done
      else
        f = File.open(instapaper_filepath, 'r', encoding: 'utf-8')
      end
    else
      warning_message_screen("No Instapaper file found")
      throw :done
    end

    begin
      f.each_line do |line|
        # Parse each line
        parts = line.split(" \\ ")
        # log_message_screen("  #{line} --> #{parts}")
        # parse the given date-time string, then create YYYYMMDD version of it
        date_YYYYMMDD = Date.parse(parts[0]).strftime('%Y%m%d')
        log_message_screen("  Found item to save with date #{date_YYYYMMDD}:")

        # Format line to add. Guard against possible empty fields
        parts[2] = '' if parts[2].nil?
        parts[3] = '' if parts[3].nil?
        line_to_add = "- #article **[#{parts[1].strip}](#{parts[2].strip})** #{parts[3].strip}"
        log_message_screen(line_to_add)

        # Read in the NP Calendar file for this date
        this_note = NPCalFile.new(date_YYYYMMDD)

        # Add new lines to end of file, creating a ### Media section before it if it doesn't exist
        this_note.append_line_to_section(line_to_add, MEDIA_STRING)
        this_note.rewrite_cal_file
        main_message_screen("-> Saved new Instapaper item to #{date_YYYYMMDD}")
      end

      unless defined?($instapaper_test_data)
        f.close
        # Now rename file to same as above but _YYYYMMDDHHMM on the end
        archive_filename = "#{IFTTT_ARCHIVE_FILEPATH}#{INSTAPAPER_FILE[0..-5]}_#{$date_time_now_file_fmttd}.txt"
        File.rename(instapaper_filepath, archive_filename)
      end
    rescue StandardError => e
      error_message_screen("ERROR: #{e.exception.message} when processing file #{INSTAPAPER_FILE}")
    end
  end
end

#--------------------------------------------------------------------------------------
# MEDIUM articles
#--------------------------------------------------------------------------------------
def process_medium
  medium_filepath = IFTTT_FILEPATH + MEDIUM_FILE
  catch (:done) do  # provide a clean way out of this
    if defined?($medium_test_data)
      f = $medium_test_data
      log_message_screen("Using Medium test data")
    elsif File.exist?(medium_filepath)
      if File.empty?(medium_filepath)
        warning_message_screen("Note: Medium file empty")
        throw :done
      else
        f = File.open(medium_filepath, 'r', encoding: 'utf-8')
      end
    else
      warning_message_screen("No Medium file found")
      throw :done
    end

    begin
      f.each_line do |line|
        # Parse each line
        parts = line.split(" \\ ")
        # log_message_screen("  #{line} --> #{parts}")
        # parse the given date-time string, then create YYYYMMDD version of it
        date_YYYYMMDD = Date.parse(parts[0]).strftime('%Y%m%d')
        log_message_screen("  Found item to save with date #{date_YYYYMMDD}:")

        # Format line to add. Guard against possible empty fields
        parts[2] = '' if parts[2].nil?
        line_to_add = "- #article **[#{parts[1].strip}](#{parts[2].strip})**"
        log_message_screen(line_to_add)

        # Read in the NP Calendar file for this date
        this_note = NPCalFile.new(date_YYYYMMDD)

        # Add new lines to end of file, creating a ### Media section before it if it doesn't exist
        this_note.append_line_to_section(line_to_add, MEDIA_STRING)
        this_note.rewrite_cal_file
        main_message_screen("-> Saved new Medium item to #{date_YYYYMMDD}")
      end

      unless defined?($medium_test_data)
        f.close
        # Now rename file to same as above but _YYYYMMDDHHMM on the end
        archive_filename = "#{IFTTT_ARCHIVE_FILEPATH}#{MEDIUM_FILE[0..-5]}_#{$date_time_now_file_fmttd}.txt"
        File.rename(medium_filepath, archive_filename)
      end
    rescue StandardError => e
      error_message_screen("ERROR: #{e.exception.message} when processing file #{MEDIUM_FILE}")
    end
  end
end

#--------------------------------------------------------------------------------------
# TWITTER
#--------------------------------------------------------------------------------------
def process_twitter
  twitter_filepath = IFTTT_FILEPATH + TWITTER_FILE
  catch (:done) do  # provide a clean way out of this
    if defined?($twitter_test_data)
      f = $twitter_test_data
      log_message_screen("Using Twitter test data")
    elsif File.exist?(twitter_filepath)
      if File.empty?(twitter_filepath)
        warning_message_screen("Note: Twitter file empty")
        throw :done
      else
        f = File.open(twitter_filepath, 'r', encoding: 'utf-8')
      end
    else
      warning_message_screen("No Twitter file found")
      throw :done
    end

    begin
      needs_concatenating = false
      previous_line = ''
      f.each_line do |line|
        # Cope with tweets over several lines: concatenate with next line
        if needs_concatenating
          line = previous_line + line
          log_message_screen("  Concatenated -> '#{line}' with next")
          needs_concatenating = false
        end
        # Parse each line
        parts = line.split(" | ")
        # If we have less than 4 parts we'll need to join this into the next line
        if parts.size < 4
          needs_concatenating = true
          previous_line = line.strip # remove whitespace (including the probable newline on the end)
          log_message_screen("  Need to concatenate '#{line}' with next")
          next
        end
        # log_message_screen("  #{line} --> #{parts}")
        # parse the given date-time string, then create YYYYMMDD version of it
        date_YYYYMMDD = Date.parse(parts[0]).strftime('%Y%m%d')

        log_message_screen("  Found item to save with date #{date_YYYYMMDD}:")

        # Format line to add
        line_to_add = "- @#{parts[2].strip} tweet: \"#{parts[1].strip}\" ([permalink](#{parts[3].strip}))"
        log_message_screen(line_to_add)

        # Read in the NP Calendar file for this date
        this_note = NPCalFile.new(date_YYYYMMDD)

        # Add new lines to end of file, creating a ### Media section before it if it doesn't exist
        this_note.append_line_to_section(line_to_add, MEDIA_STRING)
        this_note.rewrite_cal_file
        main_message_screen("-> Saved new Twitter item to #{date_YYYYMMDD}")
        needs_concatenating = false
        previous_line = ''
      end

      unless defined?($twitter_test_data)
        f.close
        # Now rename file to same as above but _YYYYMMDDHHMM on the end
        archive_filename = "#{IFTTT_ARCHIVE_FILEPATH}#{TWITTER_FILE[0..-5]}_#{$date_time_now_file_fmttd}.txt"
        File.rename(twitter_filepath, archive_filename)
      end
    rescue StandardError => e
      error_message_screen("ERROR: #{e.exception.message} when processing file #{TWITTER_FILE}")
    end
  end
end

#=======================================================================================
# Main logic
#=======================================================================================

# Setup program options
options = {}
opt_parser = OptionParser.new do |opts|
  opts.banner = "NotePlan media adder v#{VERSION}" # \nDetails at https://github.com/jgclark/NotePlan-tools/\nUsage: npMediaSave.rb [options]"
  opts.separator ''
  options[:instapaper] = false
  options[:medium] = false
  options[:spotify] = false
  options[:twitter] = false
  options[:verbose] = false
  opts.on('-h', '--help', 'Show this help') do
    puts opts
    exit
  end
  opts.on('-i', '--instapaper', 'Add Instapaper records') do
    options[:instapaper] = true
  end
  opts.on('-m', '--medium', 'Add Medium records') do
    options[:medium] = true
  end
  opts.on('-s', '--spotify', 'Add Spotify records') do
    options[:spotify] = true
  end
  opts.on('-t', '--twitter', 'Add Twitter records') do
    options[:twitter] = true
  end
  opts.on('-v', '--verbose', 'Show information as I work') do
    $verbose = true
  end
end
opt_parser.parse! # parse out options, leaving file patterns to process

log_message_screen("\nStarting npMediaSave at #{$date_time_now_log_fmttd}")
process_instapaper if options[:instapaper]
process_medium if options[:medium]
process_spotify if options[:spotify]
process_twitter if options[:twitter]
